### ISSUE TITLE
Add the capability to automatically close the MessageBox after the specific timeout.

### ISSUE THREAD

**Issue by DmitryGaravsky** [dotnet/winforms#162]
This request is based on popular SO question: [Close a MessageBox after several seconds](https://stackoverflow.com/questions/14522540/close-a-messagebox-after-several-seconds/14522952#14522952) (100+ votes).
<a href="https://stackoverflow.com/a/14522952/1010363"><img alt="StackOverflow Answer" src="https://img.shields.io/badge/StackOverflow-QnA-green.svg"></a>

### Rationale

When developing LOB application it is quite often needed to interact with the user via the MessageBox which automatically closes after the specific timeout. The standard `MessageBox.Show` API does not provide this capability. 
The new API should be based on standard Win32 MessageBox and should be compatible with corresponding .Net implementation (`System.Windows.Forms.MessageBox`).

The [AutoClosingMessageBox](https://github.com/DmitryGaravsky/AutoClosingMessageBox) repository contains a good example of WinAPI-based wrapper on standard Win32 MessageBox compatible with the .Net `System.Windows.Forms.MessageBox`. 
<a href="https://www.nuget.org/packages/AutoClosingMessageBox/"><img alt="Nuget Version" src="https://img.shields.io/nuget/v/AutoClosingMessageBox.svg" data-canonical-src="https://img.shields.io/nuget/v/AutoClosingMessageBox.svg" style="max-width:100%;" /></a>

### Usage
Here how you can use the new API:

Scenario 1: Fire and forget - it about to be closed after default timeout(1000ms)
```c#
AutoClosingMessageBox.Show("Hello, you are using the AutoClosingMessageBox first time!");
```

Scenario 2: Wait for some result or make the default decision
```c#
var result = AutoClosingMessageBox.Show(
            text: "To be or not to be?", 
            caption: "The question",
            timeout: 2500,
            buttons: MessageBoxButtons.YesNo,
            defaultResult: DialogResult.Yes);
if(result == DialogResult.Yes) {
    // to be
}
else { 
    // or not
}
```
Additionaly you can use the `AutoClosingMessageBox.Factory` method to wrap the specific MessageBox showing. This way allows us do not provide all the `AutoClosingMessageBox.Show` method overloads which associated with the corresponting `System.Windows.Forms.MessageBox.Show` methods.
```c#
var toBeOrNotToBeQuestion = AutoClosingMessageBox.Factory(
        showMethod: (caption, buttons) =>
            MessageBox.Show(this, "To be or not to be?", caption, buttons, MessageBoxIcon.Question),
        caption: "The question"
    );
if(DialogResult.Yes == toBeOrNotToBeQuestion.Show(
                            timeout: 2500,
                            buttons: MessageBoxButtons.YesNo,
                            defaultResult: DialogResult.Yes)) {
    // to be
}
else {
    // or not
}
```
### Proposed API:

```c#
public static class AutoClosingMessageBox {
    public static DialogResult Show(string text, string caption = null, int timeoutMilliseconds = 1000, MessageBoxButtons buttons = MessageBoxButtons.OK, DialogResult defaultResult = DialogResult.None);
    public static DialogResult Show(IWin32Window owner, string text, string caption = null, int timeoutMilliseconds = 1000, MessageBoxButtons buttons = MessageBoxButtons.OK, DialogResult defaultResult = DialogResult.None);
}
```

The `AutoClosingMessageBox.Factory` method (optional, but suggested):
```c#
public static class AutoClosingMessageBox {
    public static IAutoClosingMessageBox Factory(Func<string, MessageBoxButtons, DialogResult> showMethod, string caption = null);
    //
    public interface IAutoClosingMessageBox {
        DialogResult Show(int timeoutMilliseconds = 1000, MessageBoxButtons buttons = MessageBoxButtons.OK, DialogResult defaultResult = DialogResult.None);
    }
}
```

**Comment by merriemcgaw** [comment]
Thanks for the suggestion! Could you follow the [API Review Process](https://github.com/dotnet/winforms/blob/master/Documentation/issue-guide.md#high-quality-feature-and-api-suggestions) to move this along? In the meantime I will move this to the future milestone.

**Comment by sharwell** [comment]
💡 I would suggest using `TimeSpan` instead of `int` for the timeout.

**Comment by euroUK** [comment]
Why it should be separate AutoClosingMessageBox  class? Why it can't be just extra optional parameter on MessageBox calss?

Also, it looks like we require to add extra state to DialogResult to specify that mbox was closed automatically.

**Comment by DmitryGaravsky** [comment]
@sharwell 
Thanks for participating, Sam. 
Initially, the `timeout` parameter was designed according to the widely used `Task.Delay(1000)` pattern. 
Using the `TimeSpan` do not allow us to make the timeout parameter optional because we can't specify the default value `TimeSpan.FromMilliseconds(1000)` doe to in not a compile-time constant.
BTW, I have renamed the parameter `timeout` to the `timeoutMilliseconds` in the proposal.

**Comment by DmitryGaravsky** [comment]
@euroUK 
Thanks for participating, Yury.
There are **two important reasons for dot not touching the default MessageBox API**:
At first, **because of potential breaking changes**. 
From my experience, it is often needed to create reflection-based proxies for the standard MessageBox API in order to create cross-platform UI libraries or frameworks which provide some level of abstraction.
Another reason **is following the modern and simple API concepts** . 
The MessageBox API is created for the framework version which do not provide the optional arguments. As result we have now over 20 overloads for `Show` method. Extending this overloads with additional parameter requires to add yet another 20 overloads...
With separate `AutoClosingMessageBox` class (of course its final name sould be discussed) we can use simplified API for popular scenarios(`Show` method) and special API (`Factory` method) which allows to wrap any specific call of legacy `MessageBox.Show` method. You can see the example in my proposal.

Regarding the extra state for DialogResult, I do not think that is really needed because we control this result via the `defaultResult` optional parameter.

**Comment by RussKie** [comment]
We can't modify `MessageBox` API. However you can achieve auto-closing (manual or automated) functionality using [`TaskDialog` API](https://github.com/dotnet/samples/tree/master/windowsforms/TaskDialogDemo/cs).

**Comment by DmitryGaravsky** [comment]
@RussKie 
Hi Igor! Thanks for participating.

> We can't modify `MessageBox` API.

I have never talked about modifying the MessageBox API. Please read [my proposal](https://github.com/dotnet/winforms/issues/162#issue-387928034) and [arguments](https://github.com/dotnet/winforms/issues/162#issuecomment-444838920) carefully)


> you can achieve auto-closing (manual or automated) functionality

Sure, I have already achieved the required functionality years ago with using [quite simple code](https://github.com/DmitryGaravsky/AutoClosingMessageBox). 
However, it's time to [upgrade the resulting lib](https://github.com/DmitryGaravsky/AutoClosingMessageBox/commit/0997ed92f959d6b5ab7f74470214a585e993559a) onto the latest official .net core version, thanks for the reminder)

**Comment by RussKie** [comment]
> I have never talked about modifying the MessageBox API. Please read [my proposal](https://github.com/dotnet/winforms/issues/162#issue-387928034) and [arguments](https://github.com/dotnet/winforms/issues/162#issuecomment-444838920) carefully)

I appreciate it. 
However Windows Forms is a wrapper for Win32 controls to allow faster and simpler development of native Windows apps. Adding a completely custom functionality is not really on our roadmap, 3rd party libraries are better suited for that.

And by providing the Task Dialog API we have provided a way to achieve the desired functionality, this is why I closed your proposal.
